File: equipment_booking_app/utils/progress.py
import threading

_lock = threading.Lock()
_data = {
    "task": None,
    "input_path": "",
    "output_path": "",
    "total": 0,
    "completed": [],
    "pending": [],
    "percent": 0,
    "status": "idle",
}
_args = ("","")

def start(task: str, input_path: str, output_path: str, files: list[str]):
    global _data, _args
    with _lock:
        _args = (input_path, output_path)
        _data = {
            "task": task,
            "input_path": input_path,
            "output_path": output_path,
            "total": len(files),
            "completed": [],
            "pending": list(files),
            "percent": 0,
            "status": "running",
        }

def get():
    with _lock:
        return dict(_data)

def update(file_name: str):
    with _lock:
        if file_name in _data["pending"]:
            _data["pending"].remove(file_name)
            _data["completed"].append(file_name)
        if _data["total"]:
            _data["percent"] = int(len(_data["completed"]) / _data["total"] * 100)

def set_status(status: str):
    with _lock:
        _data["status"] = status

def args():
    with _lock:
        return _args

def reset():
    start(None, "", "", [])
    set_status("idle")

File: equipment_booking_app/workflow_automation/static/js/progress.js
document.addEventListener('DOMContentLoaded', function () {
  const container = document.getElementById('zip-progress-container');
  if (!container) return;
  const bar = document.getElementById('zip-progress-bar');
  const percent = document.getElementById('zip-progress-percent');
  const taskName = document.getElementById('zip-task-name');
  const inputEl = document.getElementById('zip-input-path');
  const outputEl = document.getElementById('zip-output-path');
  const doneList = document.getElementById('zip-files-completed');
  const pendingList = document.getElementById('zip-files-pending');
  
  function getCookie(name) {
    let cookieValue = null;
    if (document.cookie && document.cookie !== '') {
      const cookies = document.cookie.split(';');
      for (let i = 0; i < cookies.length; i++) {
        const cookie = cookies[i].trim();
        if (cookie.substring(0, name.length + 1) === (name + '=')) {
          cookieValue = decodeURIComponent(cookie.substring(name.length + 1));
          break;
        }
      }
    }
    return cookieValue;
  }
  const csrftoken = getCookie('csrftoken');

  function sendAction(action) {
    fetch('/progress-control/', {
      method: 'POST',
      headers: {'X-CSRFToken': csrftoken},
      body: new URLSearchParams({action})
    });
  }

  document.getElementById('pause-btn').addEventListener('click', () => sendAction('pause'));
  document.getElementById('resume-btn').addEventListener('click', () => sendAction('resume'));
  document.getElementById('cancel-btn').addEventListener('click', () => sendAction('cancel'));
  document.getElementById('restart-btn').addEventListener('click', () => sendAction('restart'));

  function update() {
    fetch('/progress-status/')
      .then(r => r.json())
      .then(data => {
        if (data.status && data.status !== 'idle') {
          container.style.display = 'block';
          taskName.textContent = data.task || 'Zipping RCP Files';
          bar.style.width = data.percent + '%';
          percent.textContent = data.percent + '%';
          inputEl.textContent = data.input_path || '';
          outputEl.textContent = data.output_path || '';

          doneList.innerHTML = '';
          (data.completed_files || []).forEach(f => {
            const li = document.createElement('li');
            li.textContent = f;
            doneList.appendChild(li);
          });

          pendingList.innerHTML = '';
          (data.pending_files || []).forEach(f => {
            const li = document.createElement('li');
            li.textContent = f;
            pendingList.appendChild(li);
          });
        } else {
          container.style.display = 'none';
        }
      });
  }

  update();
  setInterval(update, 3000);
});
